README
# FEIDIAN #

FEIDIAN 1.00 (cli)
Copyright (C) 2003-2018 Derrick Sobodash (derrick@sobodash.com)
https://gitlab.com/sobodash/feidian

## Using This File ##

I designed README.md as a case study to give the user a feel for how FEIDIAN
can be applied to real-world development. I find something like this is more
useful to a new user than a simple list of commands and syntax.

Please refer to COMMAND.md for a list of commands


## Contents ##

1.   What is FEIDIAN?
2.   Setup
3.   Monochrome bitplane graphics
4.   Creating a custom tile definition
5.   Extracting a graphic
6.   Inserting a graphic
7.   Color bitplane graphics
8.   Scaling and padding tiles
9.   Building a dual-letter font
10.  BCR (Binary Character Recognition)
11.  Other Notes


## What is FEIDIAN? ##

FEIDIAN is a tool for converting between tile-based bitmap and bitplane
graphic formats.

Planar graphics are an alternative system to the chunk-based approach used in
most modern images. Each bit of an image corresponds to one pixel on the screen.
Instead of storing all color bits together in one location, planar graphics
split them across multiple memory chips where they can be joined by dedicated
display hardware. It was the most cost effective way to show graphics on early
computers and video game hardware.

Some systems have planar images based on fixed tiles since their graphics are
directly accessed through RAM. Others store such images in tiles of any width,
height, and color depth, as they are transferred to VRAM through software.

Most planar graphic editors are GUI based and limited to certain predefined
formats. The goal of FEIDIAN was to produce a flexible tool that could be
incorporated into a build process and which could be extended by users to
quickly support exotic planar formats.


## Setup ##

FEIDIAN is easy to set up on most modern operating systems. It only requires
that the command shell have access to a PHP interpreter. If you are using OSX
or most Linux distributions, your operating system will already have access to
a PHP interpreter. For other Linux systems or FreeBSD, you should install your
relevant php-cli package.

Windows users will need to download PHP and append the path to PHP.EXE to their
system's $PATH$ variable.

Once PHP is installed, there are a few more options you can configure in
settings.php. Open the file up in your favorite text editor.

The first option is a warning and is enabled by default. This will alert you if
your dumping pattern could produce an invalid graphic. It is strongly
recommended you leave this enabled, but to disable it, just comment
out the line (put "//" at the start of the line to comment it out).

Next is your preferred graphic format. FEIDIAN can work with either BMP
(Windows Bitmap) or XPM (X Pixmap) graphic formats. BMP is enabled by default.
If you use an OS where a bitmap editor is not available, you can comment this
line out and uncomment the one below it to use XPM instead.

The next two options are the default name and copyright string of system fonts
(BDF or FD/FON) produced by FEIDIAN. You can change these to anything you like.
Just type the new text between the quotation marks, and if you want to use any
quotation marks within your text, put a "\" before them.

There's a load of options for Binary Character Recognition (BCR) as well. The
first is what character to write when a tile isn't matched. It's recommended
if matching dual byte tiles, you use a dual-byte character (or two spaces) to
preserve the columns in the matched file.

$char_bytes is the number of bytes used for the text encoding in your table
file. If you're feeding in SJIS, you'd write "2" since SJIS is two bytes. If
using Unicode, you'd enter "5", since Unicode is five bytes. All characters in
the file must follow this encoding. Do not try to double up encodings in a
file (example: SJIS, Unicode and EUC-KR), this will cause corrupt output.

The next up is hashing options for tile comparison. The shorter your hash, the
faster matching will be. The longer your hash, the less chance of mismatched
tiles. With no hash, there's no chance of errors, but PHP arrays are known to
be glitchy when containing too long an entry. It is recommended you use a hash,
since the chance of error is around 1:16,980,000 -- fairly good odds when
you're giving it less than 3000 characters. You can only declare ONE hashing
method. This means all expect the one you are using should be commented out.

If you must use no hash, there's an option for Gzipping the tiles. This is
your best option for hash-free matching. It can also be used with the hashes
themselves, though it probably would not improve performance (any time saved
by the smaller tiles would be wasted since it takes time to Gzip them). Just
uncomment the line to enable Gzipping.

The last option is for custom tiles. By uncommenting it, you can use the
high-ascii European characters in your tile definitions. This is commented out
by default, so if you do use them in a definition, alert people to uncomment
this line before using your tile definition!


## Standard bitplane font format ##

So let's take a VERY common tile size, 8x16. This means the tile is 8 bits
wide and 16 bits tall. Since a byte is 8 bits, this tile would be 16 bytes in
the file ((8*16)/8). Simple, yes? I thought so.

So far, we're still in the realm of things every other tile tool on the planet
can do -- that being multiples of 8. People like working with bytes because
they're fast and fairly regular. When you get into tiles that do not store
their rows in complete bytes, it can get a little hairy.

Say you have a file that is 12x14. That means there are 12 bits to a line. Good
luck finding a tile editor that will handle that. If you find one, send the
author a hearty thank you because virtually NOBODY fells like adding support
for such widths, even though it's INCREDIBLY common.

So how are these 12 bits being stored? Well, most commonly, multiple bytes are
just hooked together in a long, wrapping bit string. Let's pretend our first
byte is "A," our second is "B," and so on. A will be repeated 8 times for each
bit in A.

                               AAAAAAAABBBB
                               BBBBCCCCCCCC
                               DDDDDDDDEEEE
                               EEEEFFFFFFFF
                               ...

And so on. This means we have one full byte on the first line, the upper nibble
(4 bits of a byte) on that line, the lower nibble on the next line, then the
next byte. Then the pattern repeats.

-------------------------------------------------------------------------------
 iv. Creating a custom tile definition
-------------------------------------------------------------------------------
So FEIDIAN will handle the above example, since it's all one after another. But
what if you ran across a REALLY queer game that used something 12x12, but did
not even bother keeping the bits in that logical an order (*cough* NCS/Masaya's
Der Langrisser *cough*). In Der Langrisser, your bits are like this:

                               AAAAAAAABBBB
                               CCCCCCCCBBBB
                               DDDDDDDDEEEE
                               FFFFFFFFEEEE
                               ...

Well uh-oh. FEIDIAN is not a crystal ball, so it can't see every possible tile
pattern imaginable. But it DOES allow the user to create their own custom tile
patterns! This is quite possibly the most useful feature in FEIDIAN.

How does one do this? First off, download any sample template off the FEIDIAN
webpage (http://feidian.sourceforge.net/) and pop it open in Notepad, vi,
or if you're scared to use a computer, pico. This file is heavily commented
(a comment is a line beginning with //) and should give you all the info you
need. But in case it's not enough, I'll walk you through making one from
scratch.

First off, a tile definition needs to say how tall and how wide a tile is.
There are two variables in the file to do this: $tile_width and $tile_height
(in PHP, a variable always has $ before the name). So for our 12x12 tile
example, we would write:
  $tile_width  = 12;
  $tile_height = 12;
Easy! Next, we need to define out above pattern. Well let's think carefully, so
we can keep our pattern as lean as possible. In the above example, the pattern
repeats itself after byte C. So that means, we only need to define up to that.
  $plane1  = "AAAAAAAABBBB
              CCCCCCCCBBBB";
Voila. You have just defined your byte pattern. That was the hardest part! The
only thing left is to say how many bytes are needed for the pattern. In this
case, it's three bytes (A, B and C). So we'll put:
  $pat_size    = 3;

You're all done! Save your file with a meaningful name with the extension .php.
(For this example, I chose dl.php since it's Der Langrisser). Also make sure
the first line of your file is "<?" and the last is "?>" Otherwise, FEIDIAN
will crash if you try to use it.

So, easy enough so far? I hope so.

-------------------------------------------------------------------------------
 v. Extracting a font
-------------------------------------------------------------------------------
Now it's time to actually dump the font. You need to know where it begins at
in the ROM. This I cannot help you with, you'll need to either debug and find
where it's loading from, or use your eyeballs and a lot of patience! Neill
Corlett's "Nana" tile viewer can be very helpful for this, but it can't show
anything that's not a multiple of 8 in 1bpp. There's a trick though, to help
you with 12x12 tiles. Since the width is 12, the nearest multiple of 8 would
be 24. If you set the width to 24, it should help you find your tiles (though
it will look broken).

Find the offset in the file where your tiles begin and write it down. It should
be something like 0xdeadbeef (poor example, but dead beef is my favorite kind).
Armed with the offset, your tile size, or your custom definition, you are
ready to use FEIDIAN.

FEIDIAN is capable of three modes. Insertion, ripping, and creating custom
tile sets using a list of letters (more on this later!) The modes are:

  r - rips bitplane to bitmap
  i - injects bitplane to bitmap
  d - creates dual-letter set.

There's also a prefix, "c," for when you're using a custom set.

If we were going with our very first example, which was not a custom tile, we
would run FEIDIAN like this:

  feidian -r 12,12,16,16,0xdeadbeef ourrom.rom output

This is a VERY simple command line. Let's go through its components.

  width,height,rows,columns,offset

The width is your tile width, 12. The height is also 12 for this example.
The 16 is how many tiles should be in a column. The next 16 is how many rows
of tiles you want to dump. The above example would dump the first 256 12x12
tiles located at 0xdeadbeef in ourrom.rom. It would output a bitmap of them
to output_12x12.bmp.

Easy enough? Kick ass.

Now for our Der Langrisser example, it's a bit different.

  feidian -cr dl,16,16,0x11814c derlang.smc output

Notice the -cr? That means we're using a custom tile definition. Out command
line is the same as before, except now, we replace the tile width and height
with out custom definition (which we saved as dl.php). Your definition should
be in the /tiles/ folder of the FEIDIAN directory.

The offset shown is actually the real offset for Der Langrisser's 12x12 tiles.
So have you followed the walkthrough so far, this is actually a functional
command line you could use to dump that font.

-------------------------------------------------------------------------------
 vi. Inserting a font
-------------------------------------------------------------------------------
Ok. So now you're a master of making your own tile definitions and dumping
fonts with FEIDIAN. What else can it do? Well, let's take that nifty
output_12x12.bmp we just made an crack it open in Paintbrush. Flood the thing
with black and bust open the text tool (white text, black background). Make
a big textbox and type something fun, like

  I ABSOLUTELY DESPISE HOW ANNOYING
  THE STUPID SSH SYSTEM IS FOR
  UPLOADING NEW FILES.

Now, let's try this.

  feidian -ci dl,16,16,0x11814c output_12x12.bmp derlang.smc

Congratulations, you have just replaced all those first tiles in Der Langrisser
with the above text. This is totally useless, since you can't see it in the
game right now (unless you inserted a string of all those first tiles in a row),
but this shows how FEIDIAN can insert a graphic back into the game.

-------------------------------------------------------------------------------
 vii. Color tiles/Multiple planes
-------------------------------------------------------------------------------
Color is a wonderful thing. It's what allows us to have games more complicated
than TI-Calculator games and Pong. But color takes more than what default
monochrome dumping can cover. To do color, we'll need to return to custom
tiles.

Remember how we previously defined a pattern for $plane1? Well, to make colors,
we need to layer planes. We can define multiple planes up to 4 for use with
FEIDIAN.

An explanation of console graphic storage is a bit beyond the scope of this
document. If you're interested in it, I suggest you head to romhacking.com and
read Klarth's document on Console Graphics. This document will focus more on
dumping color from an end-user perspective.

FEIDIAN does not ship with color support. To do color tiles, you will need to
download the appropriate tile definitions from the FEIDIAN website and install
them in your /tiles/ folder. All formats explained in Karth's document have
already been supported, so you won't need to design them yourself.

Once you have it installed, you can find graphics using Nana or your favorite
tile editor (since Nana can't do some layerings that tile editors might), then
just dump how you normally would using a custom tile.

You may notice the colors will be VERY funky! This is not an accident. The tile
definition contains a palette in it, which you may wish to edit to get a more
accurate reflection of the colors used in your game. Because graphics are
handled by their binary data, the palette is irrelevant to reinsertion, it's
just something you can make pretty for your editing pleasure.

Each color in the palette is stored as an RGB array. The first # is red, the
second is green and the third is blue. You can change them to any value between
0-255 (of 0x00-0xff) you like.

After you have your palette the way you want, just redump the graphic.

-------------------------------------------------------------------------------
 viii. Scaling and padding
-------------------------------------------------------------------------------
Both monochrome and color bitmaps can be scaled by FEIDIAN with no problems.
To scale a graphic, use this command line:

  feidian -s *width,*height input.bmp output

The width and height are how much you wish to multiply by. You can only do
nearest neighbor scaling in increments of 100%, so a scale setting of 2,1 would
scale the graphic to 200%x100%.

Padding, likewise, can work on both monochrome and color graphics. You can pad
by x,y pixels (width,height). These are pixels IN ADDITION to the current tile
size, not a new tile size. The command line for padding is:

    feidian -p tile_width,tile_height,pad_width,pad_height input.bmp output

-------------------------------------------------------------------------------
 ix. Building a dual-letter font
-------------------------------------------------------------------------------
How would this be useful? We're getting there, hold your horses.

First off, we will need a font. A FONT YOU SAY?! Yes, but not just any font.
It needs to be half the width of our desired tile size, 16 columns wide, and 16

CHANGES
FEIDIAN CHANGE LOG

- 0.90b -
* Fixed compatibility with PHP 7.0.
* Added user-submitted tile definitions to the official repository.

- 0.90a -
* And after more errors, it turns out to be DWORD. UGH!
* Fixed an error in fontfile.php that allowed color images to be used (this
  will ALWAYS result in broken output).
* Fixed an error in bcr.php that was causing color images to read as
  monochrome, which made BCR completely break.

- 0.90 -
* FINALLY fixed the bitmap glitch. Previously MadMalkav and I thought it was
  a image width issue, hence the option in settings ot force valid width. This
  turned out to be incorrect. The bitmap specification I was working on failed
  to mention all image data has to be stored in WORDs, not BYTEs. All files
  have been updated for this fix. FEIDIAN should finally work with all possible
  widths in Bitmap and XPM formats.
* Fixed numerous spelling errors in this change log that have accumulated the
  last 8 months.

- v0.89 -
* Added support for Tile Layer Pro color palettes. This should make it easier
  to get the colors you want in FEIDIAN output. In Tile Layer Pro, you can
  drag around the sliders to change colors. When you get all the colors how
  you want, this can be saved to a .TPL. You can include these .TPL files in
  FEIDIAN by adding them AFTER the output file in -cr ripping. It is only
  supported in -cr since -cr is the only non-monochrome ripping function.

- v0.86 -
* If you need to interleave every pixel (like GBA does), just add
  "$interleave=1;" to your tile definition. This would transform 01234567 into
  10325476, which some systems do. The interleave is performed AFTER all
  planes have been merged, so if you needed to interleave only certain planes
  (which would be absolutely psychotic from a development standpoint) you
  would still need to make a custom dump/insert routine in the tile
  definition.
* Changed the copyright tag

- v0.85a -
* Fixed a glitch with tile padding which appears to have been present for
  several versions. Padding by an odd integer width will break. The reason is
  yet unknown (it SHOULD be making valid bitmap widths and appears to expand
  the tiles properly...). For now, FEIDIAN just blocks you from using odd
  integers. There were also some other fixes and speed-ups to the pad code.
* Not directly included, but there's now a tile definition for
  Game Boy Advance on the website made thanks to some information on the
  format provided by Kojiro Sasaki.

- v0.85 -
* Fixed an severe error (set $bit2 twice) that was causing all 3bpp insertion
  to break.
* Fixed a glitch with linear insertion. Genesis/Megadrive seems to work
  properly now. Can anyone verify if this holds true for NeoGeo Pocket Color?
* Added CRC32 as an option for tile data in BCR. If you want to use CRC32,
  uncomment the line in settings.php.
* Changed the MD5 and SHA1 for BCR to use raw output instead of alpha-numeric.
  This means the output will be half the length, making tile search even
  quicker.
* Optimized 60 lines in both custominj.php and customrip.php. But this is
  pointless since full XPM support required adding a ton more.
* Full support for XPM reading and writing in all functions. Hopefully this
  proves useful to someone. On the upside of things, with XPM support GDLIB 2.0
  should now be at my disposal.
* Added a line to start muting notices. There's this minor annoyance about
  declaring a color depth twice when working with custom tiles. It's because
  the tile definition gets included twice. I've updated the tile definition
  template to check if a color depth has been defined and updated all tile
  definitions which people have submitted to the website, but this is just in
  case you have some creeping around your computer based on the old format.
* 8width.php has been completely phased out.

- v0.8b -
* Copyright.php has been renamed to settings.php (since it now has more
  settings than just the default font copyright).
* BCR improved to allow you to compare tiles by MD5sum or SHA1 hashes, as
  well as gzipping the data. Options for this are in settings.php. Please read
  the notes in settings.php before attempting any kind of BCR.
* European characters at ASCII 0xa0-0xef can be used in tile definitions, IF
  you enable them. Since Asian language systems cannot process these
  characters, it is disabled by default. If you create a tile definition
  dependant on these characters, please include a note that it requires
  EXTEND_LETTERS==TRUE.
* A new value, FORCE_PROPER_WIDTH, has been added to the settings. This
  addresses an issue MadMalkav discovered where color graphics corrupt if
  (columns*tile_width)%8!=0. This only affects tile widths that are not a
  multiple of 8. If you are using such a tile width, consider dumping in a
  multiple of 8 columns (8, 16, 24, 32, etc). If you forget, FEIDIAN will print
  a notice and terminate. You can disable this in settings.php.
* Custom insert and extract broken into two separate files to cut down on the
  size and make it easier to edit (less to load, too).
* Fixed a glitch that was breaking monochrome custom definitions since color
  had been added. I don't know how the hell this slipped by testing.

- v0.8a -
* Major fix to the custom tile routine. The new hierarchy for custom tile
  letters is: [A-Z] [a-z] [2-9] [!?@*]

- v0.8 -
* COLOR! COLOR! AND THEN MORE COLOR! I finally got around to it. This means
  there are a few changes to the tile definition format. You can download any
  of the sample definitions (they've all been edited for compatibility) from
  the page and have a look, or check out the definition template that comes
  with FEIDIAN now. Currently, all formats, 16 colors or under, in Klarth's
  console graphics document are supported, as well as some others. This means:
    -   8x8 2bpp NES
    - 16x16 2bpp NES
    -   8x8 2bpp SNES/Gameboy/GBC
    -   8x8 2bpp NeoGeo Pocket Color
    -   8x8 2bpp Virtual Boy
    -   8x8 3bpp SNES

mkwinfont.py
#!/usr/bin/python

import sys
import string

# Generate Windows bitmap font files from a text description.

def byte(i):
    return chr(i & 0xFF)
def word(i):
    return byte(i) + byte(i >> 8)
def dword(i):
    return word(i) + word(i >> 16)

def frombyte(s):
    return ord(s)
def fromword(s):
    return frombyte(s[0:1]) + 256 * frombyte(s[1:2])
def fromdword(s):
    return fromword(s[0:2]) | (fromword(s[2:4]) << 16)

def asciz(s):
    i = string.find(s, "\0")
    if i != -1:
	s = s[:i]
    return s

class font:
    pass

class char:
    pass

def loadfont(file):
    "Load a font description from a text file."
    f = font()
    f.copyright = f.facename = f.height = f.ascent = None
    f.italic = f.underline = f.strikeout = 0
    f.weight = 400
    f.charset = 0
    f.pointsize = None
    f.chars = [None] * 256

    fp = open(file, "r")
    lineno = 0
    while 1:
	s = fp.readline()
	if s == "":
	    break
	lineno = lineno + 1
	while s[-1:] == "\n" or s[-1:] == "\r":
	    s = s[:-1]
	while s[0:1] == " ":
	    s = s[1:]
	if s == "" or s[0:1] == "#":
	    continue
	space = string.find(s, " ")
	if space == -1:
	    space = len(s)
	w = s[:space]
	a = s[space+1:]
	if w == "copyright":
	    if len(a) > 59:
		sys.stderr.write("Copyright too long\n")
		return None
	    f.copyright = a
	    continue
	if w == "height":
	    f.height = string.atoi(a)
	    continue
	if w == "facename":
	    f.facename = a
	    continue
	if w == "ascent":
	    f.ascent = string.atoi(a)
	    continue
	if w == "pointsize":
	    f.pointsize = string.atoi(a)
	    continue
	if w == "weight":
	    f.weight = string.atoi(a)
	    continue
	if w == "charset":
	    f.charset = string.atoi(a)
	    continue
	if w == "italic":
	    f.italic = a == "yes"
	    continue
	if w == "underline":
	    f.underline = a == "yes"
	    continue
	if w == "strikeout":
	    f.strikeout = a == "yes"
	    continue
	if w == "char":
	    c = string.atoi(a)
	    y = 0
	    f.chars[c] = char()
	    f.chars[c].width = 0
	    f.chars[c].data = [0] * f.height
	    continue
	if w == "width":
	    f.chars[c].width = string.atoi(a)
	    continue
	try:
	    value = string.atol(w, 2)
	    bits = len(w)
	    if bits < f.chars[c].width:
		value = value << (f.chars[c].width - bits)
	    elif bits > f.chars[c].width:
		value = value >> (bits - f.chars[c].width)
	    f.chars[c].data[y] = value
	    y = y + 1
	except ValueError:
	    sys.stderr.write("Unknown keyword "+w+" at line "+"%d"%lineno+"\n")
	    return None

    if f.copyright == None:
	sys.stderr.write("No font copyright specified\n")
	return None
    if f.height == None:
	sys.stderr.write("No font height specified\n")
	return None
    if f.ascent == None:
	sys.stderr.write("No font ascent specified\n")
	return None
    if f.facename == None:
	sys.stderr.write("No font face name specified\n")
	return None
    for i in range(256):
	if f.chars[i] == None:
	    sys.stderr.write("No character at position " + "%d"%i + "\n")
	    return None
    if f.pointsize == None:
	f.pointsize = f.height
    return f

def fnt(font):
    "Generate the contents of a .FNT file, given a font description."

    # Average width is defined by Windows to be the width of 'X'.
    avgwidth = font.chars[ord('X')].width
    # Max width we calculate from the font. During this loop we also
    # check if the font is fixed-pitch.
    maxwidth = 0
    fixed = 1
    for i in range(0,256):
	if avgwidth != font.chars[i].width:
	    fixed = 0
	if maxwidth < font.chars[i].width:
	    maxwidth = font.chars[i].width
    # Work out how many 8-pixel wide columns we need to represent a char.
    widthbytes = (maxwidth+7)/8
    widthbytes = (widthbytes+1) &~ 1  # round up to multiple of 2
    # widthbytes = 3 # FIXME!

    file = ""
    file = file + word(0x0300) # file version
    file = file + dword(0)     # file size (come back and fix later)
    copyright = font.copyright + ("\0" * 60)
    copyright = copyright[0:60]
    file = file + copyright
    file = file + word(0)      # font type (raster, bits in file)
    file = file + word(font.pointsize) # nominal point size
    file = file + word(100)    # vertical resolution
    file = file + word(100)    # horizontal resolution
    file = file + word(font.ascent)   # top of font <--> baseline
    file = file + word(0)      # internal leading
    file = file + word(0)      # external leading
    file = file + byte(font.italic)
    file = file + byte(font.underline)
    file = file + byte(font.strikeout)
    file = file + word(font.weight)   # 1 to 1000; 400 is normal.
    file = file + byte(font.charset)
    if fixed:
	pixwidth = avgwidth
    else:
	pixwidth = 0
    file = file + word(pixwidth) # width, or 0 if var-width
    file = file + word(font.height) # height
    if fixed:
	pitchfamily = 0
    else:
	pitchfamily = 1
    file = file + byte(pitchfamily) # pitch and family
    file = file + word(avgwidth)
    file = file + word(maxwidth)
    file = file + byte(0)          # first char
    file = file + byte(255)        # last char
    file = file + byte(0)          # default char
    file = file + byte(32)         # break char
    file = file + word(widthbytes) # dfWidthBytes
    file = file + dword(0)         # device
    file = file + dword(0)         # face name
    file = file + dword(0)         # BitsPointer (used at load time)
    file = file + dword(0)         # pointer to bitmap data
    file = file + byte(0)          # reserved
    if fixed:
	dfFlags = 1
    else:
	dfFlags = 2
    file = file + dword(dfFlags)   # dfFlags
    file = file + word(0) + word(0) + word(0) # Aspace, Bspace, Cspace
    file = file + dword(0)         # colour pointer
    file = file + ("\0" * 16)      # dfReserved1

    # Now the char table.
    offset_chartbl = len(file)
    offset_bitmaps = offset_chartbl + 257 * 6
    # Fix up the offset-to-bitmaps at 0x71.
    file = file[:0x71] + dword(offset_bitmaps) + file[0x71+4:]
    bitmaps = ""
    for i in range(0,257):
	if i < 256:
	    width = font.chars[i].width
	else:
	    width = avgwidth
	file = file + word(width)
	file = file + dword(offset_bitmaps + len(bitmaps))
	for j in range(widthbytes):
	    for k in range(font.height):
		if i < 256:
		    chardata = font.chars[i].data[k]
		else:
		    chardata = 0
		chardata = chardata << (8*widthbytes - width)
		bitmaps = bitmaps + byte(chardata >> (8*(widthbytes-j-1)))

    file = file + bitmaps
    # Now the face name. Fix up the face name offset at 0x69.
    file = file[:0x69] + dword(len(file)) + file[0x69+4:]
    file = file + font.facename + "\0"
    # And finally fix up the file size at 0x2.
    file = file[:0x2] + dword(len(file)) + file[0x2+4:]

    # Done.
    return file

def direntry(f):
    "Return the FONTDIRENTRY, given the data in a .FNT file."
    device = fromdword(f[0x65:])
    face = fromdword(f[0x69:])
    if device == 0:
	devname = ""
    else:
	devname = asciz(f[device:])
    facename = asciz(f[face:])
    return f[0:0x71] + devname + "\0" + facename + "\0"

stubcode = [
  0xBA, 0x0E, 0x00, # mov dx,0xe
  0x0E,             # push cs
  0x1F,             # pop ds
  0xB4, 0x09,       # mov ah,0x9
  0xCD, 0x21,       # int 0x21
  0xB8, 0x01, 0x4C, # mov ax,0x4c01
  0xCD, 0x21        # int 0x21
]
stubmsg = "This is not a program!\r\nFont library created by mkwinfont.\r\n"

def stub():
    "Create a small MZ executable."
    file = ""
    file = file + "MZ" + word(0) + word(0)
    file = file + word(0) # no relocations
    file = file + word(4) # 4-para header
    file = file + word(0x10) # 16 extra para for stack
    file = file + word(0xFFFF) # maximum extra paras: LOTS
    file = file + word(0) + word(0x100) # SS:SP = 0000:0100
    file = file + word(0) # no checksum
    file = file + word(0) + word(0) # CS:IP = 0:0, start at beginning
    file = file + word(0x40) # reloc table beyond hdr
    file = file + word(0) # overlay number
    file = file + 4 * word(0) # reserved
    file = file + word(0) + word(0) # OEM id and OEM info
    file = file + 10 * word(0) # reserved
    file = file + dword(0) # offset to NE header
    assert len(file) == 0x40
    for i in stubcode: file = file + byte(i)
    file = file + stubmsg + "$"
    n = len(file)
    pages = (n+511) / 512
    lastpage = n - (pages-1) * 512
    file = file[:2] + word(lastpage) + word(pages) + file[6:]
    # Now assume there will be a NE header. Create it and fix up the
    # offset to it.
    while len(file) % 16: file = file + "\0"
    file = file[:0x3C] + dword(len(file)) + file[0x40:]
    return file

def fon(name, fonts):
    "Create a .FON font library, given a bunch of .FNT file contents."

    # Construct the FONTDIR.
    fontdir = word(len(fonts))
    for i in range(len(fonts)):
	fontdir = fontdir + word(i+1)
	fontdir = fontdir + direntry(fonts[i])

    # The MZ stub.
    stubdata = stub()
    # Non-resident name table should contain a FONTRES line.
    nonres = "FONTRES 100,96,96 : " + name
    nonres = byte(len(nonres)) + nonres + "\0\0\0"
    # Resident name table should just contain a module name.
    mname = ""
    for c in name:
	if c in "0123546789ABCDEFGHIJKLMNOPQRSTUVWXYZabcdefghijklmnopqrstuvwxyz":
	    mname = mname + c
    res = byte(len(mname)) + mname + "\0\0\0"
    # Entry table / imported names table should contain a zero word.
    entry = word(0)

    # Compute length of resource table. 2 for the shift count, plus
    # 20 for each resource, plus 2 for end-of-table, plus 8 for
    # the "FONTDIR" resource name.
    # Resources are currently one FONTDIR plus n fonts. TODO: a 
    # VERSIONINFO resource would be nice too.
    resrcsize = 12 + 20 * (1 + len(fonts))
    resrcpad = ((resrcsize + 15) &~ 15) - resrcsize

    # Now position all of this after the NE header.
    p = 0x40        # NE header size
    off_segtable = off_restable = p
    p = p + resrcsize + resrcpad
    off_res = p
    p = p + len(res)
    off_modref = off_import = off_entry = p
    p = p + len(entry)
    off_nonres = p
    p = p + len(nonres)

    pad = ((p+15) &~ 15) - p
    p = p + pad
    q = p + len(stubdata)

    # Now q is file offset where the real resources begin. So we can
    # construct the actual resource table, and the resource data too.
    restable = word(4) # shift count
    resdata = ""
    # The FONTDIR resource.
    restable = restable + word(0x8007) + word(1) + dword(0)
    restable = restable + word((q+len(resdata)) >> 4)
    start = len(resdata)
    resdata = resdata + fontdir
    while len(resdata) % 16: resdata = resdata + "\0"    
    restable = restable + word((len(resdata)-start) >> 4)
    restable = restable + word(0x0C50) + word(resrcsize-8) + dword(0)
    # The font resources.
    for i in range(len(fonts)):
	restable = restable + word(0x8008) + word(1) + dword(0)
	restable = restable + word((q+len(resdata)) >> 4)
	start = len(resdata)
	resdata = resdata + fonts[i]
	while len(resdata) % 16: resdata = resdata + "\0"    
	restable = restable + word((len(resdata)-start) >> 4)
	restable = restable + word(0x1C30) + word(0x8001 + i) + dword(0)
    # The zero word.
    restable = restable + word(0)
    assert len(restable) == resrcsize - 8
    restable = restable + "\007FONTDIR"
    restable = restable + "\0" * resrcpad

    file = stubdata + "NE" + byte(5) + byte(10)
    file = file + word(off_entry) + word(len(entry))
    file = file + dword(0) # no CRC
    file = file + word(0x8308) # the Mysterious Flags
    file = file + word(0) + word(0) + word(0) # no autodata, no heap, no stk
    file = file + dword(0) + dword(0) # CS:IP == SS:SP == 0
    file = file + word(0) + word(0) # segment table len, modreftable len
    file = file + word(len(nonres))
    file = file + word(off_segtable) + word(off_restable)
    file = file + word(off_res) + word(off_modref) + word(off_import)
    file = file + dword(len(stubdata) + off_nonres)
    file = file + word(0) # no movable entries
    file = file + word(4) # seg align shift count
    file = file + word(0) # no resource segments
    file = file + byte(2) + byte(8) # target OS and more Mysterious Flags
    file = file + word(0) + word(0) + word(0) + word(0x300)

    # Now add in all the other stuff.
    file = file + restable + res + entry + nonres + "\0" * pad + resdata

    return file

outfile = None
facename = None
fonmode = 1
infiles = []
a = sys.argv[1:]
options = 1
if len(a) == 0:
    print "usage: mkwinfont [-fnt | -fon] [-o outfile] [-facename name] files"
    sys.exit(0)
while len(a) > 0:
    if a[0] == "--":
	options = 0
	a = a[1:]
    elif options and a[0][0:1] == "-":
	if a[0] == "-o":
	    try:
		outfile = a[1]
		a = a[2:]
	    except IndexError:
		sys.stderr.write("option -o requires an argument\n")
		sys.exit(1)
	elif a[0] == "-facename":
	    try:
		facename = a[1]
		a = a[2:]
	    except IndexError:
		sys.stderr.write("option -facename requires an argument\n")
		sys.exit(1)
	elif a[0] == "-fnt":
	    fonmode = 0
	    a = a[1:]
	elif a[0] == "-fon":
	    fonmode = 1
	    a = a[1:]
	else:
	    sys.stderr.write("ignoring unrecognised option "+a[0]+"\n")
	    a = a[1:]
    else:
	infiles = infiles + [a[0]]
	a = a[1:]

if len(infiles) < 0:
    sys.stderr.write("no input files specified\n")
    sys.exit(1)

if outfile == None:
    sys.stderr.write("no output file specified\n")
    sys.exit(1)

if fonmode == 0 and len(infiles) > 1:
    sys.stderr.write("FNT mode can only process one font\n")
    sys.exit(1)

fnts = []
fds = []
for fname in infiles:
    f = loadfont(fname)
    if f == None:
	sys.stderr.write("unable to load font description "+fname+"\n")
	sys.exit(1)
    fds = fds + [f]
    fnts = fnts + [fnt(f)]

if fonmode == 0:
    outfp = open(outfile, "wb")
    outfp.write(fnts[0])
    outfp.close()
else:
    # If all supplied fonts have the same face name, use that.
    # Otherwise, require that one be input.
    autoname = fds[0].facename
    for f in fds[1:]:
	if autoname != f.facename:
	    autoname = None
    if facename == None:
	facename = autoname
    if facename == None:
	sys.stderr.write("fonts disagree on face name; "+\
	"specify one with -facename\n")
	sys.exit(1)
    outfp = open(outfile, "wb")
    outfp.write(fon(facename, fnts))
    outfp.close()
